drafts/str4d-overwinter-activation/draft1.rst

```diff
@@ -0,0 +1,196 @@
+activation mechanism is necessary so that nodes can unambiguously know at what point the changes from an
+upgrade come into effect (and can enforce consensus rule changes, for example).
+
+
+Specification
+=============
+
+The following constants are defined for every network upgrade:
+
+BRANCH_ID
+  A globally-unique 32-bit identifier.
+
+ACTIVATION_HEIGHT
+  The block height at which the network upgrade rules will come into effect, and be enforced as part of the
+  blockchain consensus.
+
+  For removal of ambiguity, the block at height ``ACTIVATION_HEIGHT - 1`` is subject to the pre-upgrade
+  consensus rules, and would be the last common block in the event of a persistent pre-upgrade branch.
+
+  It MUST be greater than the value of ``DEPRECATION_HEIGHT`` in the last software version that will not
+  contain support for the network upgrade. It SHOULD be chosen to be reached approximately three months after
+  the first software version containing support for the network upgrade is released.
+
+Activation mechanism
+--------------------
+
+There MUST only be one network upgrade at a time in progress. Concretely, this means that the Zcash blockchain
+is broken into "epochs" of block height intervals ``[ACTIVATION_HEIGHT_{N-1}, ACTIVATION_HEIGHT_N)`` (ie.
+including ``ACTIVATION_HEIGHT_{N-1}`` and exluding ``ACTIVATION_HEIGHT_N``), on which consensus rule sets are
+defined.
+
+Consensus rules themselves (and any network behavior or surrounding code that depends on them) MUST be gated
+by block height checks. For example:
+
+.. code:: cpp
+
+  if (chainActive.Tip()->nHeight >= OVERWINTER.ACTIVATION_HEIGHT) {
+      // Overwinter-specific logic
+  } else {
+      // Pre-Overwinter logic
+  }
+
+  // ...
+
+  if (pindex->nHeight >= OVERWINTER.ACTIVATION_HEIGHT) {
+      // Overwinter consensus rules applied to block
+  } else {
+      // Pre-Overwinter consensus rules applied to block
+  }
+
+
+Block parsing
+`````````````
+Incoming blocks known to have a particular height (due to their parent chain being entirely known) MUST be
+parsed under the consensus rules corresponding to their height.
+
+Incoming blocks with unknown heights (because at least one block in their parent chain is unknown) MUST NOT be
+considered valid, but MAY be cached for future consideration after all their parents have been received.
+
+Chain reorganization
+````````````````````
+It is possible for a reorganization to occur that rolls back from after the activation height, to before that
+height. This can handled in the same way as any regular chain orphaning or reorganization, as long as the new
+chain is valid over the same epochs.
+
+Post-activation upgrading
+`````````````````````````
+If a user does not upgrade their node to a compatible software version before ``ACTIVATION_HEIGHT`` is
+reached, their node will follow the pre-upgrade chain and download blocks that are incompatible with the
```

Comment:
Checking that use of terms "chain" and "branch" here were intended.

Reply:
Good catch, should be "branch".